### DISCUSSION TITLE
Need help to confirm that the Apple OAuth2 provider returns the user email

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#5074] [Announcements]
While several users reported that the current Apple OAuth2 account linking works, recently there were also several reports here and to the support mail that the user email is not returned when using the Apple OAuth2 provider.

As explained in [some of the posts](https://github.com/pocketbase/pocketbase/discussions/3814#discussioncomment-7665553), I can't verify them because I don't have Apple developer account, nor access to Apple device.

Since this was reported by at least 5 different users I've decided to update slightly the current implementation and to allow support for POST redirect callback in case `response_mode=form_post` is used as it seems to be required by Apple if we wanted to specify scopes.

But again I have no way to confirm that the above changes actually works, so I'll appreciate if users that have Apple developer account could test it.

The test binaries are available as [v0.22.14-rc](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.14-rc).

The easiest way to test it is:

1. Download the [prebuilt v0.22.14-rc release](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.14-rc) for your platform.

2. Put the below plain html content in `pb_public/index.html`:
  ```html
  <!DOCTYPE html>
  <html>
  <head>
      <meta charset="utf-8" />
      <meta name="viewport" content="width=device-width, initial-scale=1" />
      <title>Apple OAuth2 test</title>
  </head>
  <body>
      <button id="btn">Login with Apple</button>
      <pre id="response"></pre>

      <script type="module">
          import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@master/dist/pocketbase.es.mjs"

          const pb = new PocketBase("http://127.0.0.1:8090");

          document.getElementById("btn").addEventListener("click", () => {
              pb.collection('users').authWithOAuth2({
                  provider: "apple",
              }).then((data) => {
                  document.getElementById("response").innerText = JSON.stringify(data, null, 2);
              }).catch((err) => {
                  console.log("Failure:", err, err.originalError);
                  document.getElementById("response").innerText = JSON.stringify(err.response, null, 2);
              });
          })
      </script>
  </body>
  </html>
  ```

 3. Enable the Apple OAuth2 provider in the PocketBase _Admin UI > Settings > Auth providers_.
 
    You can follow this [Okta tutorial](https://developer.okta.com/blog/2019/06/04/what-the-heck-is-sign-in-with-apple) to see how to create the various required provider keys. You can skip the last "Generate the Client Secret" ruby step because the Admin UI comes with a secret generator (the "Generate secret" button in the Apple provider settings panel).
    
    You'll need:
    - client_id (aka. "Apple's Services Id")
    - client_secret which could be generated from the Admin UI when provided:
      - client_id
      - team_id
      - key_id
      - private_key (not stored on the server; used only for the secret generation)

    Note that for redirect url in your Apple OAuth2 app you have to set `http://127.0.0.1:8090/api/oauth2-redirect`.
    If Apple doesn't allow plain HTTP or localhost addresses and requires valid public HTTPS url, you could use a ssh tunneling service like [ngrok](https://ngrok.com/), [zrok](https://zrok.io/), etc. to create a public address for your PocketBase server:
    ```sh
    ./pocketbase serve

    # in another terminal run
    ngrok http 8090

    # use the generated url (make sure to update also the url in the SDK constructor!)...
    ```

4. Open your browser, navigate to `http://127.0.0.1:8090` (_or the generated ngrok/zrok HTTPS address_) and click on the "Login with Apple" button.

If it fails you could also try starting the executable with `--dev` to inspect if there are more detailed error logs.

If it succeed and still doesn't return the email, please verify that in the success auth response under "rawUser" there is "email" and "email_verified" fields (the latter should be `true`/`"true"`/`1` in order for PocketBase to assign the email field).

**Comment by danohn** [comment]
Hi @ganigeorgiev, when I run this with [v0.22.14-rc](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.14-rc), the pop-up window show a 405 Method Not Allowed error. The email is also not displayed. Let me know if I can further help with any testing.

**Reply by ganigeorgiev** [reply]
Are you sure that you've setup the correct callback url (eg. `https://example.com/api/oauth2-redirect`)?

Could you start the executable with `--dev` and paste the error log request here (or check the Admin UI > Logs) to see what is the request that results in 405?

**Reply by danohn** [reply]
>Are you sure that you are sending the request to the correct callback url (eg. https://example.com/api/oauth2-redirect)?

Yes, I am sure of this. In my case, the domain is my Ngrok URL.

> Could you start the executable with --dev and paste the error log request here (or check the Admin UI > Logs) to see what is the request that results in 405?

Sure, [here are the logs for v0.22.13](https://github.com/user-attachments/files/15880813/v0.22.13_dev_logs.txt). And [here are the logs for v0.22.14-rc](https://github.com/user-attachments/files/15880838/v0.22.14-rc_dev_logs.txt).

I do note that in v0.22.14-rc the error is related to a POST to `/_/` which can't be seen in v0.22.13.

**Reply by ganigeorgiev** [reply]
> The email is also not displayed. 

Were you presented with a screenshot similar to the one from the Okta tutorial?

![share_email_screenshot](https://developer.okta.com/assets-jekyll/blog/sign-in-with-apple/apple-permission-prompt-a21fa70b97f68dfb0c169418b706921fa4dcf1444cd0cdaaba2ad4b49badcf12.png)

**Reply by ganigeorgiev** [reply]
> the error is related to a POST to /_/

Oh, I see. I'll add a test for that endpoint too but if it actually tries to redirect to the static `/_/` success/failure pages this means that the Apple redirect has been completed (aka. the 405 error in this case shouldn't affect the auth completion).

**Reply by danohn** [reply]
>Were you presented with a screenshot similar to the one from the Okta tutorial?

Yes, I do see the email on the consent page from Apple:
<img width="828" alt="constent" src="https://github.com/pocketbase/pocketbase/assets/82357071/9b818da1-c39a-4539-95c0-2faf70337ad3">

However, the JSON payload that is displayed in the `<pre>` is missing the email:

```
{
  "meta": {
    "id": "1.2.3",
    "name": "",
    "username": "",
    "email": "",
    "avatarUrl": "",
    "accessToken": "a.b.c.d",
    "refreshToken": "e.f.g.h",
    "expiry": "2024-06-18 05:31:11.347Z",
    "rawUser": {
      "at_hash": "012345",
      "aud": "id.example",
      "auth_time": 1718685068,
      "exp": 1718771471,
      "iat": 1718685071,
      "iss": "https://appleid.apple.com",
      "nonce_supported": true,
      "sub": "x.y.x"
    },
    "isNew": true
  },
  "record": {
    "avatar": "",
    "collectionId": "_pb_users_auth_",
    "collectionName": "users",
    "created": "2024-06-18 04:31:11.813Z",
    "email": "",
    "emailVisibility": false,
    "id": "abcdefg",
    "name": "",
    "updated": "2024-06-18 04:31:11.813Z",
    "username": "users82473",
    "verified": true
  },
  "token": "eyABCDEF"
}
```

**Reply by ganigeorgiev** [reply]
Regarding the missing email, if you are not seeing the above popup, could you try to revoke your app access from the Apple dashboard to simulate "first-time" login (it is possible that the login will not shown if you already had previously granted access to the app with the older v0.22.13 version).

**Reply by danohn** [reply]
That worked! I revoked access to the app and then can see that the consent screen matches the one from the Okta tutorial:

<img width="1005" alt="consent_after_revocation" src="https://github.com/pocketbase/pocketbase/assets/82357071/a4b3b2d8-9009-472b-b409-0c4a05185502">

The JSON payload now includes the email:

```
{
  "meta": {
    "id": "1.2.3",
    "name": "",
    "username": "",
    "email": "i***@***.*z",
    "avatarUrl": "",
    "accessToken": "a.b.c.d",
    "refreshToken": "e.f.g.h",
    "expiry": "2024-06-18 06:39:56.785Z",
    "rawUser": {
      "at_hash": "012345",
      "aud": "id.example",
      "auth_time": 1718689193,
      "email": "i***@***.*z",
      "email_verified": true,
      "exp": 1718775597,
      "iat": 1718689197,
      "iss": "https://appleid.apple.com",
      "nonce_supported": true,
      "sub": "x.y.x"
    },
    "isNew": false
  },
  "record": {
    "avatar": "",
    "collectionId": "_pb_users_auth_",
    "collectionName": "users",
    "created": "2024-06-18 04:31:11.813Z",
    "email": "i***@***.*z",
    "emailVisibility": false,
    "id": "hqge2oin0zpyyy3",
    "name": "",
    "updated": "2024-06-18 05:39:57.024Z",
    "username": "users82473",
    "verified": true
  },
  "token": "eyABCDEF"
}
```

**Reply by ganigeorgiev** [reply]
In that case only the `/_/` redirect seems that will have to be updated to accept POST.
I'll submit the change a little bit later today. Thank you for testing.

**Reply by danohn** [reply]
You're welcome!

**Reply by jainmohit** [reply]
> Regarding the missing email, if you are not seeing the above popup, could you try to revoke your app access from the Apple dashboard to simulate "first-time" login (it is possible that the login will not shown if you already had previously granted access to the app with the older v0.22.13 version).

How to do that i cannot see an option to do that as such i tried with creating a new apple id for that its asking me share the email which i did which is working fine on browser on macbook in web browser as well the simulator app but the issue appears when login using the iphone app and giving me error attached screen shot.
![Screenshot 2024-06-25 at 12 28 32](https://github.com/pocketbase/pocketbase/assets/2365183/1d4deced-e4aa-45ed-b996-b389ffd47174)

**Reply by jainmohit** [reply]
i ran this command below to check the error in browser but i can't see any error other than the message:

https://appleid.apple.com/auth/authorize?response_type=code&redirect_uri=<redirect_uri>&client_id=<client_id>

its redirecting me to auth/oauth2-redirect-failure

**Reply by jainmohit** [reply]
i followed this documentation to setup everything:

https://medium.com/identity-beyond-borders/how-to-configure-sign-in-with-apple-77c61e336003

**Comment by Xerossd** [comment]
Thanks 👍

**Reply by lukepighetti** [reply]
Very exciting and timely for me. I'm just about to turn on SIWA and will need emails! 🎉

**Comment by ganigeorgiev** [comment]
The above 405 error for the `/_/` route is resolved in `master` and it is released as [v0.22.14](https://github.com/pocketbase/pocketbase/releases) together with a small change to the email linking addressing a security concern that was raised earlier today.

**Reply by jainmohit** [reply]
Hi there,

Even after using the latest one i am still getting the following error to login into the platform:

data
: 
{email: {code: "validation_required", message: "Cannot be blank."}}

using the apple oAuth 2 automated code.

Can you please suggest.

**Reply by ganigeorgiev** [reply]
@jainmohit Please check the comment in https://github.com/pocketbase/pocketbase/discussions/5074#discussioncomment-9801780.

Additionally, from the Apple [docs](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_rest_api/authenticating_users_with_sign_in_with_apple):

> The email address is either the user’s real email address or the proxy address, depending on their private email relay service. **This value may be empty for Sign in with Apple at Work & School users. For example, younger students may not have an email address.**

If you are still not able to make it work, outline the steps how did you create your Apple OAuth2 app so eventually other users that have Apple developer account could assist you.

**Comment by effdotsh** [comment]
@ganigeorgiev for me email works but the name field is blank
```json
"OAuth2User": {
    "id": "0**9",
    "name": "",
    "username": "",
    "email": "e**@icloud.com",
    "avatarUrl": "",
    "accessToken": "a**w",
    "refreshToken": "r**A",
    "expiry": "2024-08-12 18:22:56.870Z",
    "rawUser": {
      "at_hash": "_**w",
      "aud": "sh.eff.g**h",
      "auth_time": 1723483371,
      "email": "e**@icloud.com",
      "email_verified": true,
      "exp": 1723569776,
      "iat": 1723483376,
      "iss": "https://appleid.apple.com",
      "nonce_supported": true,
      "sub": "0**9"
    }
  },
  "IsNewRecord": false
}
```

**Reply by ganigeorgiev** [reply]
If this is subsequent "Sign-in with Apple" call then the name might not be returned. Based on my understanding of the [Apple docs](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_rest_api/authenticating_users_with_sign_in_with_apple#3383768) it will be returned **only** for the initial/first sign-in request:

> Although Apple provides the user’s email address in the identity token on all subsequent API responses, it doesn’t include other information about the user, such as their name.

To verify this you can try to revoke your app access to simulate "first-time login approval". See also the user provided screenshot in https://github.com/pocketbase/pocketbase/discussions/5074#discussioncomment-9801780 confirming that the user is successfully asked to allow access to the name and email fields.

Alternatively you could try to experiment with requesting different scopes. By default we request only `name email`. You can pass custom scopes with the JS SDK request by specifying the `scopes` array:

```js
pb.collection('users').authWithOAuth2({
    provider: "apple",
    scopes: ["name", "email", ... something else ...]
})
```

But again, note that scope changes most likely will have effect only on first login.

More than this I don't know how to assist further.

**Reply by effdotsh** [reply]
Thanks for the reply. I tested with a new account/first authentication and it still was not returning the name.  In the comment you linked, the JSON response also doesn't include the name (the website sucessfuly prompts for the name for me as well). Looking very briefly at the code, [here](https://github.com/pocketbase/pocketbase/blob/238eddc4e028ce748e088d405987febda371aa02/tools/auth/apple.go#L66) seems to be parsing the name as a string, but the [apple docs](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/configuring_your_webpage_for_sign_in_with_apple#3331292) have the response listed in the form of `{ "name": { "firstName": string, "lastName": string }, "email": string }` under a `user` item. Am I missreading the parsing (I'm on a plane right now so couldn't look too in depth) or could this be the issue?

Edit: Fixed apple docs link

**Reply by ganigeorgiev** [reply]
Your link to the "apple docs" seems to point to the PocketBase implementation and I can't find what actually you are referring to.

But in any case if the name was returned as object then it would have failed the unmarshaling. Note that it is also missing from the `rawUser` data where no extra parsing is done.

If you think that there is an issue in the PocketBase implementation feel free to open a PR with the suggested change but I'm not willing to invest more time on this.

**Reply by ganigeorgiev** [reply]
I see that you've updated the Apple docs link. Then in this case if I'm reading correctly the docs, the name fields will be returned under the `user` object and not as part of the `id_token` (_which is against the [examples in the spec](https://openid.net/specs/openid-connect-core-1_0.html#id_tokenExample)_).
If that's the case then we can simply add and extra parsing for it in https://github.com/pocketbase/pocketbase/blob/238eddc4e028ce748e088d405987febda371aa02/tools/auth/apple.go#L95.

I'll consider updating the implementation with the upcoming v0.23.0 in the next following weeks (I have no way to verify that this would work but it shouldn't hurt having the extra `user` json check).

**Reply by effdotsh** [reply]
Thank you! Gotta love apple "thinking different" haha

**Reply by ganigeorgiev** [reply]
I've added the `user` field parsing as minor release in [v0.22.19](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.19).

If that doesn't work then I'm out of ideas. Again, I have no Apple developer account to test the integration, so feel free to open a PR if other changes are needed.

**Reply by effdotsh** [reply]
Wow thank you for how fast you moved on this. Unfortunately, I was not expecting this speed so I already replaced this flow with a native implementation in my app 😅. I'm under a really tight deadline right now, but once that passes (~1 month) I'll take a look at this change and open a PR as needed. Thank you for your work on this project!

**Reply by HEAVYPOLY** [reply]
On pocketbase v0.22.21
With Apple sign in, email works but not getting name, how do I parse the apple user field?

```js
{
  "meta": {
    "id": "00**41",
    "name": "",
    "username": "",
    "email": "a*@heavypoly.com",
    "avatarUrl": "",
    "accessToken": "ab**ho4Q",
    "refreshToken": "rf11**SQ",
    "expiry": "2024-09-27 19:43:09.187Z",
    "rawUser": {
      "at_hash": "cL**9ag",
      "aud": "com.HP.HP",
      "auth_time": 1727462587,
      "email": "a*@heavypoly.com",
      "email_verified": true,
      "exp": 1727548988,
      "iat": 1727462588,
      "iss": "https://appleid.apple.com",
      "nonce_supported": true,
      "sub": "0**41"
    },
    "isNew": true
  },
  "record": {
    "avatar": "",
    "canvasCount": 0,
    "collectionId": "_pb_users_auth_",
    "collectionName": "users",
    "created": "2024-09-27 18:43:09.558Z",
    "email": "a*@heavypoly.com",
    "emailVisibility": false,
    "favorites": null,
    "hotkeys": null,
    "id": "fl**8",
    "name": "",
    "prefs": null,
    "pro": false,
    "stats": null,
    "unlimited": false,
    "unlocked": false,
    "updated": "2024-09-27 18:43:09.558Z",
    "username": "users98116",
    "variant": "",
    "verified": true
  },
  "token": "e**4"
}
```

**Reply by ganigeorgiev** [reply]
@HEAVYPOLY I have no way to verify it as I don't have Apple developer account to test the integration. Please note that, based on the [Apple docs](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/configuring_your_webpage_for_sign_in_with_apple#3331292), Apple returns the user object that contains the name **only the first time** the user authorizes the app.
